[RISCV] Copy AVLs whose LiveIntervals aren't extendable in insertVSETVLI

lukel97 opened the pull request:
Currently before forwarding an AVL we do a simple non-exhaustive check to see if its LiveInterval is extendable. But we also need to check for this when we're extending an AVL's LiveInterval via merging the VSETVLIInfos in transferBefore with equally zero AVLs.

Rather than trying to conservatively prevent these cases, this inserts a copy of the AVL instead if we don't know we'll be able to extend it. This is likely to be more robust, and even if the extra copy is undesirable these cases should be rare in practice.

lukel97 committed:
[RISCV] Copy AVLs whose LiveIntervals aren't extendable in insertVSETVLI

Currently we do a simple non-exhaustive check to see if a LiveInterval is extendable before forwarding an AVL. But we also need to check for this when we're extending the live range via merging the VSETVLIInfos in transferBefore with equally zero AVLs.

Rather than trying to conservatively prevent these cases, this inserts a copy of the AVL instead if we don't know we'll be able to extend it. This is likely to be more robust, and even if the extra copy is undesirable these cases should be rare in practice.

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1149,15 +1143,26 @@ void RISCVInsertVSETVLI::insertVSETVLI(MachineBasicBlock &MBB,
+      else {
+        Register AVLCopyReg =
+            MRI->createVirtualRegister(&RISCV::GPRNoX0RegClass);
+        MachineBasicBlock::iterator AVLDef =
```
Do you know how a PHIDef is handled here?  Is the slot index simply mapped to the first instruction in the block?  Or something else?  Either way, we may need special handling for that case here. 

For correctness, I'd appreciate if you did a local patch which removes the previous extend check and ensures everything works.  The fallback path should be generic.

lukel97 replied:
A PHI def's SlotIndex will be at the entry to the block, so it will insert a copy at the entry, which I believe is what we want.

>  I'd appreciate if you did a local patch which removes the previous extend check and ensures everything works

If I'm understanding you correctly, the previous extend check is removed in this patch already

preames replied:
> A PHI def's SlotIndex will be at the entry to the block, so it will insert a copy at the entry, which I believe is what we want.

Are we guaranteed that the Slot_Block entry maps to an MachineInstr?  Getting a MachineInstr can return nullptr, and I'm not clear about the exact circumstances.

> > I'd appreciate if you did a local patch which removes the previous extend check and ensures everything works
> 
> If I'm understanding you correctly, the previous extend check is removed in this patch already

You have:
```
      if (!LI.liveAt(SI) && LI.containsOneValue())
        LIS->extendToIndices(LI, SI);
```

I'm suggesting run "ninja check" (and other tests of your choice) with:

```
      if (false && !LI.liveAt(SI) && LI.containsOneValue())
        LIS->extendToIndices(LI, SI);
```

i.e. Force *everything* done the COPY path.  This will hurt code quality, but should never violate invariants.  (This is only a local to you test to make sure it works.)

lukel97 replied:
> Are we guaranteed that the Slot_Block entry maps to an MachineInstr? Getting a MachineInstr can return nullptr, and I'm not clear about the exact circumstances.

Thanks for catching that, turns out`getInstructionFromIndex` does return null for that. I've added explicit handling for this to insert it with the other PHIs. We also didn't have any test cases that hit this with a PHI so I've added one. 

Sorry I see what you mean about removing the extend check now, I got it confused with the check on line 953. I can confirm that does work and is generic.

lukel97 committed:
Fix PHI case handling where there is no instruction, add test case

lukel97 committed:
Don't advance phi insertion point: we want to insert before the first non-phi

As an aside, the first non phi should just be the first instruction since we're past phi elimination...

preames reviewed (approved):
LGTM

lukel97 committed:
Update lsr-drop-solution.ll

The vsetvli here is removed because we're relaxing the only-one-valno constraint for AVL forwarding
